src/Microsoft.ServiceFabric.Services.Queryable/LINQ/ReliableIndexedDictionaryQueryContext.cs

```diff
@@ -0,0 +1,70 @@
+﻿using Microsoft.Data.OData.Query;
+using Microsoft.ServiceFabric.Data;
+using Microsoft.ServiceFabric.Data.Indexing.Persistent;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Linq.Expressions;
+using System.Reflection;
+using System.Threading;
+using System.Threading.Tasks;
+using static Microsoft.ServiceFabric.Services.Queryable.IEnumerableUtility;
+
+namespace Microsoft.ServiceFabric.Services.Queryable.LINQ
+{
+    internal class ReliableIndexedDictionaryQueryContext
+    {
+        // Executes the expression tree that is passed to it. 
+        // Should return 
+        internal static object Execute<TKey, TValue, TResult>(Expression expression, IReliableStateManager stateManager, IReliableIndexedDictionary<TKey, TValue> dictionary)
+                        where TKey : IComparable<TKey>, IEquatable<TKey>
+        {
+            // The expression must represent a query over the data source. 
+            if (!IsQueryOverDataSource(expression))
+                throw new InvalidProgramException("No query over the data source was specified.");
+
+            // Find the call to Where() and get the lambda expression predicate.
+            InnermostWhereFinder whereFinder = new InnermostWhereFinder();
+            MethodCallExpression whereExpression = whereFinder.GetInnermostWhere(expression);
+            LambdaExpression lambdaExpression = (LambdaExpression)((UnaryExpression)(whereExpression.Arguments[1])).Operand;
+
+            // Send the lambda expression through the partial evaluator.
+            lambdaExpression = (LambdaExpression)Evaluator.PartialEval(lambdaExpression);
+
+            WhereExtractor<TResult> extractor = new WhereExtractor<TResult>(expression);
+            BinaryOperatorKind operatorKind = extractor.OperatorKind;
+            object constant = extractor.Constant;
+            string propertyName = extractor.PropertyName;
+
+            MethodInfo filterHelper = typeof(ReliableStateExtensions).GetMethod("FilterHelper", BindingFlags.Static | BindingFlags.Public);
+            filterHelper = filterHelper.MakeGenericMethod(new Type[] { typeof(TKey), typeof(TValue), constant.GetType() });
+            Task<IEnumerable<TKey>> keysTask = (Task<IEnumerable<TKey>>)filterHelper.Invoke(null, new object[] { dictionary, constant, operatorKind, false, new CancellationToken(), stateManager, propertyName});
+            IEnumerable<TKey> keys = keysTask.Result;
+
+            IEnumerable<TValue> values;
+            using (var tx = stateManager.CreateTransaction())
+            {
+                IEnumerable<KeyValuePair<TKey, TValue>> pairs = dictionary.GetAllAsync(tx, keys, TimeSpan.FromSeconds(4), new CancellationToken()).Result;
+                values = new KeyValueToValueEnumerable<TKey, TValue>(pairs);
+                tx.CommitAsync();
```

Comment:
Does this not cause problems?  Commit is async, but you don't await it and the transaction gets disposed immediately.  I suppose since it's just a READ, there's no replication so commit doesn't do anything?  

Also, I know Service Fabric has had issues with doing blocking calls (calling .Result in dictionary.GetAllAsync().Result) in the past.  Something to watch out for.